From cc986cf067166cd02d65cda188b87414ed9893df Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Date: Thu, 9 May 2024 16:59:54 +0530
Subject: [PATCH] wifi: ath12k: Validate data length before copying data
When the driver repeatedly requests a file in a short period from
firmware through firmware_request_nowarn() function, that is not
present in the firmware path.
In this case, due to the presence of fw_priv in the cache within
firmware_request_nowarn(), it considers the request as a batch
request and return a file of zero size without any error.
This leads to failure of phy bring-up.
To resolve this issue add a check for valid data length.
This issue can be seen  when module reboot with mlo_capable=0
cold_boot_cal=0.
Patch-dependency: 000-c-Revert-wifi-ath12k-remove-cold_boot_cal-module-param.patch
Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)
diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index a35431f..b3ce950 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -739,8 +739,12 @@ int ath12k_core_fetch_board_data_api_1(struct ath12k_base *ab,
 	if (IS_ERR(bd->fw))
 		return PTR_ERR(bd->fw);
 
-	bd->data = bd->fw->data;
-	bd->len = bd->fw->size;
+	if (bd->fw->size) {
+		bd->data = bd->fw->data;
+		bd->len = bd->fw->size;
+	} else {
+		return -ENOENT;
+	}
 
 	return 0;
 }
-- 
2.34.1